File: shared/src/test/java/ru/privatenull/pnauth/security/TotpServiceTest.java
package ru.privatenull.pnauth.security;

import org.junit.jupiter.api.Test;
import ru.privatenull.pnauth.storage.AuthRecord;
import ru.privatenull.pnauth.storage.AuthRepository;

import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class TotpServiceTest {
    @Test
    void encryptsSecretsAndCreatesProvisioningData() {
        byte[] key = new byte[32];
        TotpService service = new TotpService(new EmptyRepository(), key);
        String secret = service.generateSecret();
        String encrypted = service.encrypt(secret);

        assertEquals(secret, service.decrypt(encrypted));
        assertTrue(service.provisioningUri("Server", "Steve", secret).startsWith("otpauth://totp/"));
        assertEquals(20, service.generateRecoveryCodes(20).size());
    }

    private static final class EmptyRepository implements AuthRepository {
        @Override
        public Optional<AuthRecord> findByUniqueId(UUID uniqueId) {
            return Optional.empty();
        }

        @Override
        public Optional<AuthRecord> findByUsername(String username) {
            return Optional.empty();
        }

        @Override
        public boolean create(AuthRecord record) {
            return false;
        }

        @Override
        public void updatePassword(UUID uniqueId, ru.privatenull.pnauth.storage.PasswordHash passwordHash) {
        }

        @Override
        public void updateLastLogin(UUID uniqueId, long timestamp) {
        }

        @Override
        public boolean updateUsername(UUID uniqueId, String username) {
            return false;
        }
    }
}

File: shared/src/test/kotlin/ru/privatenull/pnauth/security/TotpServiceTest.kt
package ru.privatenull.pnauth.security

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import ru.privatenull.pnauth.storage.AuthRecord
import ru.privatenull.pnauth.storage.AuthRepository
import java.util.Optional
import java.util.UUID

class TotpServiceTest {
    @Test
    fun encryptsSecretsAndCreatesProvisioningData() {
        val key = ByteArray(32)
        val service = TotpService(EmptyRepository(), key)
        val secret = service.generateSecret()
        val encrypted = service.encrypt(secret)

        assertEquals(secret, service.decrypt(encrypted))
        assertTrue(service.provisioningUri("Server", "Steve", secret).startsWith("otpauth://totp/"))
        assertEquals(20, service.generateRecoveryCodes(20).size)
    }

    private class EmptyRepository : AuthRepository {
        override fun findByUniqueId(uniqueId: UUID): Optional<AuthRecord> = Optional.empty()
        override fun findByUsername(username: String): Optional<AuthRecord> = Optional.empty()
        override fun create(record: AuthRecord): Boolean = false
        override fun updatePassword(uniqueId: UUID, passwordHash: ru.privatenull.pnauth.storage.PasswordHash) {}
        override fun updateLastLogin(uniqueId: UUID, timestamp: Long) {}
        override fun updateUsername(uniqueId: UUID, username: String): Boolean = false
    }
}
